samples/startup-script/index.js

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright 2018, Google, Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const Compute = require('@google-cloud/compute');
+const http = require('http');
+
+const compute = new Compute();
+
+const zone = compute.zone('us-central1-a');
+
+// callback(error, externalIp)
+function createVm(name, callback) {
+  // Create a new VM, using default ubuntu image. The startup script
+  // installs apache and a custom homepage.
+  const config = {
+    os: 'ubuntu',
+    http: true,
+    metadata: {
+      items: [
+        {
+          key: 'startup-script',
+          value: `#! /bin/bash
+
+        # Installs apache and a custom homepage
+        apt-get update
+        apt-get install -y apache2
+        cat <<EOF > /var/www/html/index.html
+        <html><body><h1>Hello World</h1>
+          <p>This page was created from a simple start up script!</p>
+        </body></html>`,
+        },
+      ],
+    },
+  };
+
+  zone
+    .createVM(name, config)
+    .then(data => {
+      const vm = data[0];
+      const operation = data[1];
+
+      operation.on('complete', () => {
```

Comment:
I believe you can return the operation and getMetadata call here, which should help keep the promise chain flat to catch errors:

```js
const vm = zone.vm(name);

vm
  .create(config)
  .then(data => {
    const operation = data[1];
    return operation.promise();
  })
  .then(() => {
    return vm.getMetadata();
  })
  .then(data => {
    const metadata = data[0];

    // ...
```

Reply:
Flattened. Thanks, looks much better this way.